src/storage/src/source/postgres.rs

```diff
@@ -718,19 +745,45 @@ impl PostgresTaskInfo {
         tokio::pin!(stream);
 
         let mut last_data_message = Instant::now();
-        let mut last_feedback = Instant::now();
         let mut inserts = vec![];
         let mut deletes = vec![];
 
-        while let Some(item) = stream.next().await {
+        let mut last_feedback = Instant::now();
+        let mut committed_lsn: Option<PgLsn> = None;
+
+        loop {
+            let data_next = stream.next();
+            tokio::pin!(data_next);
+            let offset_recv = self.offset_rx.recv();
+            tokio::pin!(offset_recv);
+
+            use futures::future::Either;
+            let item = match futures::future::select(offset_recv, data_next).await {
+                Either::Left((Some(to_commit), _)) => {
+                    // We assume there is only a single partition here.
+                    let lsn: PgLsn = to_commit[&PartitionId::None].offset.into();
+                    // Set the committed lsn so we can send a correct status update
+                    // next time we are required.
+                    committed_lsn = Some(lsn);
+                    continue;
+                }
+                Either::Right((Some(item), _)) => item,
+                Either::Left((None, _)) | Either::Right((None, _)) => {
+                    break;
+                }
+            };
+
             let item = item?;
             use ReplicationMessage::*;
 
             // The upstream will periodically request status updates by setting the keepalive's
             // reply field to 1. However, we cannot rely on these messages arriving on time. For
             // example, when the upstream is sending a big transaction its keepalive messages are
-            // queued and can be delayed arbitrarily. Therefore, we also make sure to send a
-            // proactive status update every 30 seconds.
+            // queued and can be delayed arbitrarily. Therefore, we also make sure to
+            // send a proactive status update every 30 seconds, but only after we receive
+            // resumption_frontier advancement. There is an implicit requirement that
+            // a new resumption frontier is converted into an lsn within 30 seconds of
```

Comment:
30 seconds from receiving the frontier. In reality i think it could be up to 1 min of delay, but that should be fine, and we can play with it as we see issues. Ill adjust the comment.

Also, snapshots dont interact with status updates at all, they only matter during replication!